### ISSUE TITLE
Cross-platform core

### ISSUE THREAD

**Issue by et1975** [fable-elmish/elmish#116]
There are now 2 non-fable builds of the same elmish core (that I know of) and I wonder if it's worth trying to accommodate both platforms and maybe make 2 different packages from the same codebase: netstandard and fable.

**Comment by ctaggart** [comment]
@et1975 What are the other two non-fable builds that you know of. This is one:
https://github.com/fable-websockets/Fable.Websockets.Elmish/blob/master/src/Fable.Websockets.Elmish/src/Elmish.fs

I really enjoyed this video and I've watch it a couple of time.
https://youtu.be/1dJ-TlBr4Ro

Do you think it would be possible to combine Elmish with zeromq, perhaps with the pure .NET https://github.com/zeromq/netmq? I'm simply confused about the Elmish dispatcher loop and other loops and how they would interact. For example zeromq has a thread based [poller loop](http://learning-0mq-with-pyzmq.readthedocs.io/en/latest/pyzmq/multisocket/zmqpoller.html) and you can subscribe to events. How would we go about getting those events on to the dispatcher. In a code base I'm working on, we have an F# mailbox processor loop listening the netmq socket. Same question. I'm a bit lost in how these loops interact.

But yes, I think a netstand2.0 package would be great that doesn't require Fable.

**Comment by et1975** [comment]
Yeah, sorry I haven't been able to dedicate any time to this.
@ctaggart [WPF](prolucid/Elmish.WPF) and [Xamarin](csoren/elmish-xamarin-forms). Some of my colleagues have done an internal project with Elmish.WPF and 0mq, not sure about the specifics, maybe @2sComplement can chime in.
I've been looking at various co-op async IO loops (dotnetty, tokio-rs, LMAX disruptor, etc)  and I can certainly see the appeal, but to me they have slightly different semantics/goals/priorities so I wouldn't try to integrate the two loops too deeply and just keep them running side-by-side and pass the messages in-between as needed. Elmish has a couple of options to `Sub` to external source of events, `Cmd.ofSub` and `Program.withSubscription` - the former can be issued from inside the `init` or `update` and your handler will get the up-to-date model, the latter will be called only once, with the initial model - otherwise the two are the same. How you'd publish the events from Elmish into those other loops will depend on the specifics obviously, but from Elmish perspective - you'd make the publish function available to the `update` as a command or as a function.

**Comment by 2sComplement** [comment]
A subscription can be set up fairly easily using the methods that @et1975 mentioned. In one Elmish.WPF project, I set up a subscription loop when the socket is connected using something like this:
```fsharp
let update msg model =
    match msg with
    | Connected zmq -> 
        let sub dispatch = 
            let callbacks = [] //some functions I want the subscription loop to call
            ZmqClient.subscribe zmq callbacks
        model, Cmd.ofSub sub
```

**Comment by ctaggart** [comment]
I don't understand fully yet, but I'll have a look at:
- https://github.com/Prolucid/Elmish.WPF
- https://github.com/csoren/elmish-xamarin-forms

And may be put together something for review.

**Comment by pirrmann** [comment]
Would the idea be to have an Elmish core with the same concepts, but that doesn't depend on Fable, and can be resused in several technologies? I'd quite like that and would be ready to work on it if you think it's relevant.

**Comment by et1975** [comment]
@pirrmann That's the idea. Realistically though, I think it will have to be (two) core builds with different package IDs, as 
- Fable needs sources and
- Might be a good idea to keep .NET or Fable packaging decoupled... cause who knows?

**Comment by et1975** [comment]
Released in 2.0 beta
